mithril-signer/src/runtime/runner.rs

```diff
@@ -262,77 +255,53 @@ impl Runner for SignerRunner {
         Ok(())
     }
 
-    async fn can_i_sign(&self, pending_certificate: &CertificatePending) -> StdResult<bool> {
-        debug!("RUNNER: can_i_sign");
-        if self
+    async fn can_sign_current_epoch(&self) -> StdResult<bool> {
+        let epoch = self
             .services
-            .signed_entity_type_lock
-            .is_locked(&pending_certificate.signed_entity_type)
+            .epoch_service
+            .read()
             .await
-        {
-            debug!(" > signed entity type is locked, can NOT sign");
-            return Ok(false);
-        }
-
-        let current_signer: Option<Signer> = {
-            let current_signers = self.get_current_signers().await;
-            if let Ok(signers) = current_signers {
-                signers
-                    .iter()
-                    .find(|s| s.party_id == self.services.single_signer.get_party_id())
-                    .cloned()
-            } else {
-                warn!(" > could not get current signers with stake, can NOT sign");
-                return Ok(false);
-            }
-        };
+            .epoch_of_current_data()?;
 
-        if let Some(signer) = current_signer {
-            debug!(" > got a Signer from pending certificate");
+        if let Some(protocol_initializer) = self
+            .services
+            .protocol_initializer_store
+            .get_protocol_initializer(epoch.offset_to_signer_retrieval_epoch()?)
+            .await?
+        {
+            debug!(" > got protocol initializer for this epoch ({epoch})");
 
-            if let Some(protocol_initializer) = self
+            return self
```

Comment:
It could be more readable using retrieving epoch_service before.
```
            let epoch_service = self.epoch_service_read().await;

            return epoch_service.can_signer_sign_current_epoch(
                self.services.single_signer.get_party_id(),
                protocol_initializer,
            );
```

Reply:
Done ✅